_integration-tests/validator/validator.go

```diff
@@ -0,0 +1,472 @@
+			closestLen   int
+			closestDiffs []string
+		)
+
+		for _, f := range fkvs {
+			var d diff
+			compare(v, f, "", 0, &d)
+			if d.Len() == 0 {
+				continue valid
+			}
+			if closestLen == 0 || d.Len() < closestLen {
+				closestLen = d.Len()
+				closest = [][]kv{f}
+				closestDiffs = []string{d.String()}
+			} else if closestLen == d.Len() {
+				closest = append(closest, f)
+				closestDiffs = append(closestDiffs, d.String())
+			}
+		}
+		fmt.Fprintf(&failed, "Failed to validate Trace:\n")
+		DumpKV(&failed, v)
+		fmt.Fprintf(&failed, "\nClosest traces(%d) from fake agent:", len(closest))
+		for i := 0; i < len(closest); i++ {
+			fmt.Fprintf(&failed, "\n\n#################### Candidate %d ####################\n", i)
+			DumpKV(&failed, closest[i])
+			fmt.Fprintf(&failed, "\nDiff between valid and candidate %d:\n%s", i, closestDiffs[i])
+		}
+	}
+	faileds := failed.String()
+	if len(faileds) != 0 {
+		fmt.Printf("################################################################################\nFAILED:\n%s\n", failed.String())
+		os.Exit(1)
+	}
+}
+
+// A diff keeps track of the differences between two things.
+type diff struct {
+	lines []string
+}
+
+// Len returns the number of differences added with AddDiffLine
+func (d *diff) Len() int {
+	return len(d.lines)
+}
+
+// AddDifference adds a "difference" to the diff. The arguments are
+// as those to printf, and the format string, f, should end in a newline.
+func (d *diff) AddDifference(f string, args ...interface{}) {
+	s := fmt.Sprintf(f, args...)
+	d.lines = append(d.lines, s)
+}
+
+// String assembles the differences into a single string that can
+// be printed for human consumption.
+// It is important to end differences registered with AddDifference
+// with newlines so that they are printed correctly.
+func (d *diff) String() string {
+	var b strings.Builder
+	for _, l := range d.lines {
+		io.WriteString(&b, l)
+	}
+	return b.String()
+}
+
+// Compare walks through valid and fake, creating a sort of diff between the
+// two. Unlike a normal diff, only the things in valid are expected to be in
+// fake, and fake can (and almost always does) contain more fields than valid.
+//
+// The fields are called valid and fake, because the valid argument is meant
+// to be sourced from a validation.json file, whereas the fake argument should
+// come from the fake agent. Given output from the fake agent, this is a way
+// to easily compare the output to the expected output described in a
+// validation.json file.
+//
+// For a given valid[vi], we advance through fake with fi, trying to match
+// the field's key. Once we find the valid[vi] key, we compare the values
+// including recursively for meta, metrics, and _children.
+//
+// If valid[vi] is not found, we note that it is missing and start
+// looking for valid[vi+1] from fake[lastfi], where lastfi is the
+// index of the last valid field we found in fake.
+func compare(valid, fake []kv, prefix string, indent int, d *diff) {
```

Comment:
This is a useful but also a good amount of code to maintain! Have you considered using https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff ?

Reply:
I think we could manipulate the code to use go-cmp, but we'd still have to do all of the work to restructure the traces, and then prune them. In other words, we might be able to get rid of the compare function, but the rest (and some additional code) would still be required.